sklearn/ensemble/_hist_gradient_boosting/splitting.pyx

```diff
@@ -519,21 +553,40 @@ cdef class Splitter:
 
             gain = _split_gain(sum_gradient_left, sum_hessian_left,
                                sum_gradient_right, sum_hessian_right,
-                               negative_loss_current_node,
+                               loss_current_node,
+                               monotonic_cst,
+                               lower_bound,
+                               upper_bound,
                                self.l2_regularization)
 
-            if gain > split_info.gain and gain > self.min_gain_to_split:
-                split_info.gain = gain
-                split_info.feature_idx = feature_idx
-                split_info.bin_idx = bin_idx
-                # we scan from left to right so missing values go to the right
-                split_info.missing_go_to_left = False
-                split_info.sum_gradient_left = sum_gradient_left
-                split_info.sum_gradient_right = sum_gradient_right
-                split_info.sum_hessian_left = sum_hessian_left
-                split_info.sum_hessian_right = sum_hessian_right
-                split_info.n_samples_left = n_samples_left
-                split_info.n_samples_right = n_samples_right
+            if gain > best_gain and gain > self.min_gain_to_split:
+                found_better_split = True
+                best_gain = gain
+                best_bin_idx = bin_idx
+                best_sum_gradient_left = sum_gradient_left
+                best_sum_hessian_left = sum_hessian_left
+                best_n_samples_left = n_samples_left
+
+        if found_better_split:
+            split_info.gain = best_gain
+            split_info.bin_idx = best_bin_idx
+            # we scan from left to right so missing values go to the right
+            split_info.missing_go_to_left = False
+            split_info.sum_gradient_left = best_sum_gradient_left
+            split_info.sum_gradient_right = sum_gradients - best_sum_gradient_left
+            split_info.sum_hessian_left = best_sum_hessian_left
+            split_info.sum_hessian_right = sum_hessians - best_sum_hessian_left
+            split_info.n_samples_left = best_n_samples_left
+            split_info.n_samples_right = n_samples - best_n_samples_left
+
+            # We recompute best values here but it's cheap
+            split_info.value_left = compute_value(
+                split_info.sum_gradient_left, split_info.sum_hessian_left,
+                lower_bound, upper_bound, self.l2_regularization)
+
+            split_info.value_right = compute_value(
+                split_info.sum_gradient_right, split_info.sum_hessian_right,
+                lower_bound, upper_bound, self.l2_regularization)
```

Comment:
For reviewers: apart from the leaf value computation (which is new), the logic here is actually unchanged.

It's mostly just a small optimization: instead of setting the all the attribute of the split_info at each iteration, we only do it once at the end